### PR TITLE
Support multi-instance configuration as a legacy feature

### PR DESCRIPTION
**Pull request by tmasternak**
Those are changes based on @justabitofcode multi-instance spike. I did following things:
 * Re-based it over newest develop
 * Changed naming to reflect that this is `legacy` mode of operation
 * Split legacy mode code into dedicated classes in `Legacy.MultiInstance` folder

PoA:
 * [x] Add Acceptance Tests for Legacy Multi-Instance feature

Connects to: #139

### CODE REVIEW

**Comment by tmasternak** [comment]
This one is can continue when https://github.com/Particular/NServiceBus.SqlServer/pull/162 is ready

**Comment by tmasternak** [comment]
@justabitofcode please review. I've moved code into separate classes because otherwise sql connection management in normal mode of operation and in legacy multi-instance mode made `MessagePump`, `MessageDispatcher` and `Receiver` too complex.

File: src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        public class Sender : EndpointConfigurationBuilder
+        {
+            public Sender()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.UseTransport<SqlServerTransport>()
+                        .UseSpecificSchema(queueName => queueName.Contains("Receiver") ? "receiver" : "sender")
+                        .EnableLagacyMultiInstanceMode(async address =>
```
Comment (weralabaj) [inline src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance.cs]: I think calling it "Legacy" gives the impression that you don't need to do anything. This is more like backwards compatibility, but handled in a new way. Would calling it "MultiInstanceCompatMode" or something like that be more precise?

File: src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance_with_transaction_mode_different_from_distributed.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿namespace NServiceBus.SqlServer.AcceptanceTests.LegacyMultiInstance
+{
+    using System;
+    using System.Data.SqlClient;
+    using System.Threading.Tasks;
+    using NServiceBus.AcceptanceTesting;
+    using NServiceBus.Transports.SQLServer;
+    using NUnit.Framework;
+
+    public class When_using_legacy_multiinstance_with_transaction_mode_different_from_distributed : When_using_legacy_multiinstance
```
Comment (weralabaj) [inline src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance_with_transaction_mode_different_from_distributed.cs]: Is it consistent with behaviour in v2?

Comment (tmasternak) [inline src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance_with_transaction_mode_different_from_distributed.cs]: No it's not. But multiinstance in v2 with Outbox in not working as expected so we decided to drop it in v3. @mauroservienti can you confirm?

Comment (weralabaj) [inline src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance_with_transaction_mode_different_from_distributed.cs]: Ok we need to update doco then.

File: src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance.cs (hunk 1 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer.AcceptanceTests/LegacyMultiInstance/When_using_legacy_multiinstance.cs]: The reason why "Legacy" has been used is to indicate that we users should avoid it as there are better ways to do it and that in v4 we will not support it any more. I'm not sure "Legacy" is the best word here though :/.

File: src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacyQueueCreator.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
+﻿namespace NServiceBus.Transports.SQLServer.LegacyCallbacks.MultiInstance
+{
+    using System.Data;
+    using System.Data.SqlClient;
+    using System.Threading.Tasks;
+
+    class LegacyQueueCreator : ICreateQueues
+    {
+        public LegacyQueueCreator(LegacySqlConnectionFactory connectionFactory, QueueAddressParser addressParser)
```
Comment (weralabaj) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacyQueueCreator.cs]: Comment regarding naming - as above, to me calling it "Legacy" is misleading. Maybe just "MultiInstance..."?

Comment (mauroservienti) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacyQueueCreator.cs]: It is defined in the `MultiInstance` namespace, isn't that enough?

File: src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlServerTransportInfrastructure.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            TimeSpan waitTimeCircuitBreaker;
+            if (!settings.TryGet(SettingsKeys.TimeToWaitBeforeTriggering, out waitTimeCircuitBreaker))
+            {
+                waitTimeCircuitBreaker = TimeSpan.FromSeconds(30);
+            }
+
+            Func<TransportTransactionMode, ReceiveStrategy> receiveStrategyFactory =
+                guarantee =>
+                {
+                    //TODO: there needs to be a better way of doing this
```
Comment (weralabaj) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlServerTransportInfrastructure.cs]: After a moment thinking I can't come up with anything that seems much better. Remove the comment?

File: src/NServiceBus.SqlServer/Receiving/MessagePump.cs [hunk 5]
```diff
@@ -29,14 +30,14 @@ public MessagePump(Func<TransportTransactionMode, ReceiveStrategy> receiveStrate
             inputQueue = new TableBasedQueue(addressParser.Parse(settings.InputQueue));
             errorQueue = new TableBasedQueue(addressParser.Parse(settings.ErrorQueue));
 
+            //TODO: there has to be a better place for this
```
Comment (weralabaj) [inline src/NServiceBus.SqlServer/Receiving/MessagePump.cs]: ?

**Comment by weralabaj** [comment]
OK, no more comments, but @justabitofcode should have a close look too.

File: src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs [hunk 6]
```diff
@@ -82,5 +82,20 @@ public static TransportExtensions<SqlServerTransport> TransactionScopeOptions(th
             transportExtensions.GetSettings().Set<SqlScopeOptions>(new SqlScopeOptions(timeout, isolationLevel));
             return transportExtensions;
         }
+
+        /// <summary>
+        /// Specifies connectionString lookup to be used by sql transport.
```
Comment (justabitofcode) [inline src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs]: Should we have an extra line of summary to indicate that this is for backwards compatibility and point customers to the new version?

File: src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacyQueueCreator.cs (hunk 3 above)
Comment (justabitofcode) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacyQueueCreator.cs]: Should the "legacy methods" have an Obsolete tag? Presumably we'd remove them in V4.

Comment (weralabaj) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacyQueueCreator.cs]: @mauroservienti that was more regarding using the word "Legacy". Maybe it's just me, if you think it's ok and precise enough then ignore.

File: src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlServerTransportInfrastructure.cs (hunk 4 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlServerTransportInfrastructure.cs]: :+1:

File: src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs (hunk 6 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs]: :+1:

File: src/NServiceBus.SqlServer/Receiving/MessagePump.cs (hunk 5 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Receiving/MessagePump.cs]: fixed

**Pull request merged**